keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,205 @@
+
+## Motivation
+
+The `node-role.kubernetes.io/master` (and the broader `node-role.kubernetes.io` namespace for labels) was introduced to provide a simple organizational and grouping convention for cluster users. The labels were reserved solely for organizing nodes via a convention that tools could recognize to display information to end users, and for use by opinionated external tooling that wished to simplify conventions for end users. Use of the label by components within the Kubernetes project (those projects subject to API review) was expressely forbidden. Specifically, no project could mandate the use of those labels in a conformant distribution, since we anticipated that many deployments of Kubernetes would have more nuanced control-plane topologies than simply "a control plane node".
+
+Over time, several changes to Kubernetes core and related projects were introduced that depended on the `node-role.kubernetes.io/master` label to vary their behavior in contravention to the guidance the label was approved under. This was unintentional and due to unclear reviewer guidelines that have since been more strictly enforced. Likewise, the complexity of Kubernetes deployments has increased and the simplistic mapping of control plane concepts to a node has proven to limit the ability of conformant Kubernetes distributions to self-host, as anticipated.
+
+
+### Goals
+
+This KEP:
+
+* Clarifies that the use of the `node-role.kubernetes/*` label namespace is reserved solely for end-user and external Kubernetes consumers, and:
+  * May not be used to vary behavior within Kubernetes projects that are subject to API review (kubernetes/kubernetes and all components that expose APIs under the `*.k8s.io` namespace)
+  * Is not required to be used for a cluster to be conformant
+* Describes the locations within Kubernetes that must be changed to use an alternative mechanism for behavior
+  * Suggests approaches for each
+* Describes the timeframe and migration process for Kubernetes distributions and deployments to update labels
+
+
+## Proposal
+
+### Use of `node-role.kubernetes.io/*` labels
+
+* Kubernetes components MUST NOT set or alter behavior on any label within the `node-role.kubernetes.io/*` namespace.
+* Kubernetes components (such as `kubectl`) MAY simplify the display of `node-role.kubernetes.io/*` labels to convey the node roles of a node
+* Kubernetes examples and documentation MUST NOT leverage the node-role labels for node placement
+* External users, administrators, conformant Kubernetes distributions, and extensions MAY use `node-role.kubernetes.io/*` without reservation
+  * Extensions are recommended not to vary behavior based on node-role, but MAY do so as they wish
+* First party components like `kubeadm` MAY use node-roles to simplify their own deployment mechanisms.
+* Conformance tests MUST NOT depend on the node-role labels in any fashion
```

Comment:
would be good to describe how we will prevent accidental uses from getting reintroduced in examples/docs/conformance/e2e/controllers, etc

Reply:
Added a small section, needs some thought.